### DISCUSSION TITLE
Wishlist overhaul

### DISCUSSION THREAD

**Discussion by bitesh** [nicotine-plus/nicotine-plus#1604] [Ideas]
soulseek has a feature on wishlists called that (same as title) which discards previous results deemed irrelevant by you and won't display those.
in nicotine+, I get the same results over and over, even though I already they're not what I want.
maybe, implement something like that? I can't find anything like that as is... I'm new though, so, maybe I'm just not seeing it...

that and the "open downloads folder" but I can select a file for that, so, less bothersome.

**Comment by andrewjmetzger** [comment]
Very good points @bitesh. To add to this, I would also like to see a better way to search for a specific format. For instance, if I have an album in MP3, but I want the lossless FLAC version, I can use search filters to see results for only the FLAC filetype. Currently, wishlists do not respect filters, so any automatic download is not likely to be of the desired format.

**Comment by bitesh** [comment]
also, if possible, save the position and length of my columns in search.

**Reply by slook** [reply]
I agree that it would be useful to find out why remembering columns is not robust, but this is entirely unrelated to the topic of your OP, please post about this in https://github.com/nicotine-plus/nicotine-plus/issues/1491 ...

_"If a user wishes to refine the customization of their column widths and order, then they are only able to do so reliably when only one search tab is open, because closing the tabs in the incorrect order will wipe out the changed width+order settings, with only the last open tab's settings remembered, and it's too easy to forget what's what so it's basically a lottery."_

**Reply by bitesh** [reply]
Yeh, true, but I just figured I'd do another suggestion along with my other suggestions rather than create a new thread.

**Reply by slook** [reply]
That's okay, but we prefer to keep things on topic otherwise it's impossible for the devs to keep track of items that require attention, things can get lost easily because this is a large repo.

Always check the [main tracker](https://github.com/nicotine-plus/nicotine-plus/issues) to see if there is already an open issue to avoid creating duplicate requests.

**Reply by mathiascode** [reply]
https://github.com/nicotine-plus/nicotine-plus/commit/d93da0de765065b55ec89ff5244cb07981e8d883 should make column saving more robust.

**Comment by slook** [comment]
> I get the same results over and over, even though I already they're not what I want.

This is my experience also, infact I would say that it puts me off using the Wishlist feature because it becomes annoying very quickly.

> a feature... which discards previous results deemed irrelevant

This sounds interesting, but I imagine it would be a large feature as probably another database is needed to keep track of this. I am not familiar with the workings in the Qt client but your suggestion warrants further investigation, I can understand why the feature you have suggested would be desirable.

> Currently, wishlists do not respect filters

Indeed they do not, I agree it is difficult to get relevant results from Wishlist items. Again this would need something like the aforementiond database to keep track of this, and drastic changes to the Wishlist and Result Filter mechanisms before this can be considered.

**Reply by 82d** [reply]
> Currently, wishlists do not respect filters

I came to the method long ago by adding filtering filters:
` what to find  -mp3 -mp4 -jpg -jpeg -png -mxm -ini  -sfv -nfo -txt -epub -tqd `
![Снимок экрана_2021-10-13_08-00-55](https://user-images.githubusercontent.com/15121782/137049103-af9f7bd0-67fe-4142-860f-48aac3e51192.png)

**Reply by andrewjmetzger** [reply]
Thanks @82d, this is an interesting method, and a decent workaround for the current releases of N+. Still, I worry that it seems clunky and difficult to maintain at scale. Good to know there is limited inbuilt support via search syntax, but a proper filterline (like normal search results have) would be a much better solution, long-term.

**Reply by ProFete** [reply]
> > Currently, wishlists do not respect filters
> 
> I came to the method long ago by adding filtering filters: `what to find -mp3 -mp4 -jpg -jpeg -png -mxm -ini -sfv -nfo -txt -epub -tqd` ![Снимок экрана_2021-10-13_08-00-55](https://user-images.githubusercontent.com/15121782/137049103-af9f7bd0-67fe-4142-860f-48aac3e51192.png)

is it possible to add filter result like bitrate, duration, size ? if yes where i can find the list of all working filters please ?

**Reply by slook** [reply]
You can't have filters on saved wishlist items because that ability is not implemented yet.

The normal search result filters are located at the top right of the result list on the Search Files tab, they are unique for each search. There are also default result filters which you can set on the Searches page of the Preferences dialog.

**Reply by ProFete** [reply]
do you know if these filters will be impelemented in the whishlist ? with these filters + an auto download function nictoine+ could work for us when we are not on the PC, for me 90% of the time, i always miss rares files because users disconnected

**Comment by mathiascode** [comment]
Add added date-time column for wishlist items https://github.com/nicotine-plus/nicotine-plus/issues/1369

**Comment by slook** [comment]
Add Wish including parenthesis (symbols and punctuation) is possible from within the Wishlist dialog, yet it would seem that such items do not yield any results and no new Search Files Wish tab is ever created for them.

It makes me wonder if perhaps new Wishlist items should go through the `PUNCTUATION` stripping procedure the same as new ordinary searches, or maybe it's better if they are stored as is but then stripped upon submitting each Wish search event as is done with `translate(self.translatepunctuation)` I don't know what unforeseen consequences of doing this might be.

**Reply by mathiascode** [reply]
This has since been fixed. Search terms are stripped before they're sent to the server.

**Comment by slook** [comment]
After testing recently added Length filter #1892 we should consider what is the best way of how to go about integrating the filters into the Wishlist searches.

I suggest starting with thinking about moving the `[server][autosearch]` setting from `config` into a new `wishlist.json` file?

**Comment by mathiascode** [comment]
Add optional notifications (similar to chat messages) for wishlist searches.

**Reply by slook** [reply]
As well your good idea of as user notifications, a pluginsystem notification event would be useful incase an Auto-Wish Downloader plugin might be desirable for https://github.com/nicotine-plus/nicotine-plus/issues/1955

**Comment by GoldenGlimpse** [comment]
I totally upvote for a "Discard Repeated Results" option.

**Reply by slook** [reply]
Since the result filters inlcude/exclude fields now allow usernames to be entered, this means when the planned integration of the result filters into wishlist items, then it will be possible to Discard Repeated Results in the same way as the official client does (whish is afaik by excluding usernames), plus the ability to exclude by other attributes such as file type, bitrate, size or duration, etc.

It is unknown at this time as to how such an option would be best presented.

**Comment by cproensa** [comment]
SoulseekQT has a flag for each wishlist item for active/inactive status.
I'd like to see it here too.
I know, it means to redesign the config json, but is my _wishlist's wish_.

An example use case for it:
When i get a result from the autosearch, add the files to the download queue, but those wont start right away, becasue of queue position or user is offline. In that case i disable the wish item so that:
- No new searches will appear becasue i dont need them (already queued)
- If the offline user never appears again, i can re-enable the wish
- When the download is finished, i then delete the wish item. Or periodically, I delete any wish item which is inactive and is not stalled in the download queue.

Otherwise:
If i had to delete the wish item, and the user is never to be seen again: i have to recreate a wish item (wich i may not remember exactly if the search string is not obvious). Or, i probably cant distinguish an old offline item in the queue, which originated from the wishlist (important), from another one that was a direct manual search (probably less important)

If i dont delete the wish item, while the download is queued: then if the download is finished, and removed from the queue (program restart, auto clean, etc), I may not be aware to go to the wishlist to remove it afterwards, and may download duplicate in the future. Or if mixed with many other downloads, i can't distinguish it from other manually-searched items.

**Reply by slook** [reply]
> it means to redesign the config json

My suggestion is to re-purpose the `config` file's `autosearch`  key to enable the wish items, then store extended filtering and stats data (including data about disabled wishes) in a separate wishlist.json file in a structure such as proposed in https://github.com/nicotine-plus/nicotine-plus/issues/1887#issuecomment-1038621078:

_config file (contains only enabled wish keys):_
`autosearch = ['example query 2', 'example query 3', 'example query 4']`

_wishlist.json file (contains all wish keys):_
> ```
> 'wish': "example query 1", 'item' [
>     'autosearch': False,
>     'date_added': 012346,
>     'filters': [
>         'in': "",
>         'out': "",
>         'size': "",
>         'br': "",
>         'slot': False,
>         'cc': "",
>         'type': "",
>         'length': ""
>     ],
>     'auto_download': [
>        'enabled' True,
>        'maximum': 10
>     ],
>     'notify': True,
>     'found': 0,
> ]
> 'wish': "example query 2", [
> ...
> ```

This would be a good approach because the config file needs to remain somewhat backwards compatible, the Wishlist list can be populated from the items `wishlist.json` file, with a column for Enabled checkbox set by `autosearch` key.

A redundant duplicate `autosearch` item in the json would restore a lost config or for portability between N+ profiles.

In this way, if one were to downgrade to an earlier version of Nicotine+ then the disabled wish items would disappear, but would re-appear again upon upgrading assuming the json file isn't deleted from the user data folder.

**Reply by mathiascode** [reply]
I don't think we want to keep `autosearch` in the config file, a wishlist.json file is enough.

**Comment by mathiascode** [comment]
TODO-list:

- [ ] Implement `wishlist.json` file and add corresponding Python dictionary to core
- [ ] Add 'enabled' column to wishlist UI
- [ ] Add 'date added' column to wishlist UI
- [ ] Implement new edit dialog for wishes
- [ ] Migrate `autosearch` config entry to wishlist.json

wishlist.json format (dumped directly from a Python dictionary):
```json
{
	"example query 1": {
		"enabled": false,
		"discard_repeated": false,
		"date_added": 12346,
		"filters": {
			"in": "",
			"out": "",
			"size": "",
			"br": "",
			"slot": false,
			"cc": "",
			"type": "",
			"length": ""
		}
	},
	"example query 2": {}
}
```

More options can be added later, as long as someone requests them and they make sense.

https://github.com/nicotine-plus/nicotine-plus/issues/1955 will not be implemented yet.

**Reply by slook** [reply]
> Implement new edit dialog for wishes

Is an edit dialog really necessary, given that presumably the widgets in the Result Filters bar will be recalled, populated and referenced to get and set filter values to store in the first place, and the remaining Enabled boolean will be an inline checkbox column, then why is the current Edit text entry dialog insufficient?

> "discard_repeated"

What is the intended purpose or mechanism of this parameter? I am guessing it will append all the `|usernames` in the currently open results list into `[filters][out]`. That seems like it will be a one-shot function such as a tab context menu item, in which case what does setting this boolean do?

What about enabling and disabling application of the "filters" values, I suggest adding a parameter for this named `"filtered"`.

**Reply by mathiascode** [reply]
> > Implement new edit dialog for wishes
> 
> Is an edit dialog really necessary, given that presumably the widgets in the Result Filters bar will be recalled, populated and referenced to get and set filter values to store in the first place, and the remaining Enabled boolean will be an inline checkbox column, then why is the current Edit text entry dialog insufficient?

I had something similar to the search preferences page in mind for the filters, a separation between the default filters and filter bar. With only the filter bar present, I'm slightly concerned that people will brick their wishlist searches by applying filters, realizing there are no files they want, and closing the tab before reverting the filter. 

It would also be nice to easily modify filters of any search term present in the wishlist.

> > "discard_repeated"
> 
> What is the intended purpose or mechanism of this parameter? I am guessing it will append all the `|usernames` in the currently open results list into `[filters][out]`. That seems like it will be a one-shot function such as a tab context menu item, in which case what does setting this boolean do?

We would probably add a new field to store usernames, to avoid cluttering the filter. It currently works the same as SoulseekQt, i.e. a boolean option that starts logging usernames when enabled, and resets the list when disabled.

The question is if this option should be global or not. In SoulseekQt, it can be toggled for individual wishlist items, but I'm not sure if people want to use it like this.

> What about enabling and disabling application of the "filters" values, I suggest adding a parameter for this named `"filtered"`.

Sounds good.

**Reply by slook** [reply]
> a boolean option that starts logging usernames when enabled, and resets the list when disabled... The question is if this option should be global or not.

Definitely not global, field to store usernames is needed for each individual wishlist item.

**Reply by mathiascode** [reply]
The usernames will obviously not be global, I was thinking about the discard_repeated boolean.

**Reply by mathiascode** [reply]
My updated thoughts on storage for the new wishlist features: https://github.com/nicotine-plus/nicotine-plus/issues/1437#issuecomment-2870198917

We would add a json file that stores search result data, and use the existing Nicotine+ config for wish filters and dates added instead.

**Reply by mathiascode** [reply]
On second though, there are people with massive wishlists, so probably better to just retire the `autosearch` config field and move everything wishlist-related to a separate file.

**Comment by AverageHoarder** [comment]
Excuse me if this has alreadby been suggested.

For people with large wishlists, a feature to save the current position within the wishlist would be neat.
Over the years my father has accumulated >6000 wish list searches which never fully run through since the host is restarted or the program updated before it reaches the end and starts over again.

Storing the last searched wishlist entry persistently so the search can resume at that point after a restart would be really helpful.

**Reply by mathiascode** [reply]
The wishlist should already behave this way.

**Reply by AverageHoarder** [reply]
Ah, good. I probably caused it to restart the search from scratch when I edited the config file while the container was down to make all wishlist searches lowercase then. I dislike having 2 alphabets in the wishlist window (one for uppercase and one for lowercase).

**Comment by mathiascode** [comment]
Having thought about this some more, I've come up with the following:
- Always discard previously seen results to begin with, see how it works in practice and if anyone wants an option to disable this
- Results are only considered seen if an unread wishlist tab is opened, which avoids the problem where we lose results we've never looked at
- Add a context menu option for wishes to reset the previously seen results, in order to get all results again
- Add an "Edit filters" item that opens a wishlist tab and sends a search request, so you have some results to look at when constructing filters
- Should we have a button to remember applied filters? In case someone is just filtering some wishlist results, but doesn't want the filters to apply later.

**Comment by fortunexbt** [comment]
Has there been any progress on `wishlist.json` or proper support for bulk wishlists?

I'm trying to track over **4000+ lossless (FLAC) tracks**, and currently:

* The `autosearch = [...]` config field breaks at scale (even <100 entries)
* Nicotine+ **silently resets** the config if it's too long or malformed
* There's **no UI import**, JSON-based alternative, or CLI hook

Basically, **automated high-volume tracking is impossible**, even though the underlying use case is extremely common (e.g. collectors, archivists, DJs, etc.).

Would love to see:

1. `wishlist.json` with real filters + `enabled` flags
2. GUI/CLI import/export
3. Per-wishlist-item memory of seen usernames (like SoulseekQt)
4. Optional auto-download rules

I’m happy to test alpha builds or help shape the data format.
Let me know where/how to contribute.

— fortunexbt

**Comment by mathiascode** [comment]
Christmas came early: https://github.com/nicotine-plus/nicotine-plus/pull/3561

For people brave enough to test the wishlist overhaul, try the latest development build (back up your config, just in case): https://nicotine-plus.org/doc/TESTING.html

**Comment by mathiascode** [comment]
Latest build for testing this is now available: https://nicotine-plus.org/doc/TESTING.html
